Fixed #32933 -- Documented BoundField.initial as preferred over Form.get_initial_for_field(). (#14651)
django/django: docs/ref/forms/api.txt

.. method:: Form.get_initial_for_field(field, field_name)

> Returns the initial data for a form field. The value is retrieved from

Reviewer: Maybe value -> data (a value wouldn't be evaluated?)

> :attr:`Form.initial`, or :attr:`Field.initial` if not present, and then

Reviewer: Maybe chop 'then'
Maybe chop 'if not present'

Author: Thanks, @smithdc1. On this one, I think it's important to say when `Field.initial` is used. Do you have alternative wording that includes that information? On the subjective points, I'll wait until @carltongibson weighs in before updating the PR.

Reviewer: Yes, finding the _right_ words is tricky.  

The other thought I've just had is that we should bear in mind the work of the translators. If we can't find something that is much better then maybe we should not change this paragraph? 

Anyhow, my attempt would be something like.

It tries to retrieve the data firstly from form.initial and secondly from field.initial.

Author: On this, I decided to go with the approach used in CPython's [`dict.get()`](https://docs.python.org/3/library/stdtypes.html#dict.get) docs (substituting "else" -> "otherwise"). I still think the conditional aspect is important to communicate (`Field.initial` is consulted only _if_ it's not present in `Form.initial`), which is something the `dict.get()` docs do.

Reviewer: The new wording is great. 👍

Author: Thanks, @smithdc1! 🙏

evaluated if a callable.

It is recommended to use :attr:`BoundField.initial` over
:meth:`~Form.get_initial_for_field()` because ``BoundField.initial`` has a
simpler interface. Also, unlike :meth:`~Form.get_initial_for_field()`,

> :attr:`BoundField.initial` caches its values. This is useful especially when

Reviewer: Maybe 'especially useful'. Appreciate this is marginal! 🙂

Author: After thinking more, this changes the meaning from what I was intending, IMO. Putting it before makes it read more like "very useful," but I was after especially's other meaning of "in particular." (It's not that it's _very_ useful, but rather that the usefulness is apparent more in these callable cases.)

.. method:: Form.get_initial_for_field(field, field_name)

While :meth:`~Form.get_initial_for_field()` exists, it is better to use

> :attr:`BoundField.initial` to retrieve the initial data for a form field.

Author: I'm not sure about this phrasing. I think it's better if the method is left in for people looking for it / expecting to find it, but at the same time, I think we want to steer people away from it.

Reviewer 2: Hmmm. I'd likely just add the note below

```
:meth:`~Form.get_initial_for_field()` allows you to retrieve initial data for a
form field. It retrieves data from :attr:`Form.initial` and
:attr:`Field.initial`, in that order, and evaluates any callable initial
values.
  
Note that :attr:`BoundField.initial` is preferred, as it provides simpler access to the 
initial data for a form field.
```

🤔

Author: Okay, thanks, @carltongibson. I was more concerned about the potential buggy behavior of going through `get_initial_for_field()`, but I guess that can be fixed. I'm filing an issue about it now (alluded to in my ticket)...

Reviewer 2: Happy to take suggestions on the `as it provides simpler access to the initial data for a form field.` — less might be more: `Note that :attr:`BoundField.initial` is preferred` and leave it at that? 🤔

I guess we could de-doc it, but there'll be a whole 15 years worth of people looking for it. …

.. method:: Form.get_initial_for_field(field, field_name)

> Returns the initial data for a form field. The data is retrieved from
> :attr:`Form.initial` if present, otherwise :attr:`Field.initial`, and
> evaluated if a callable.

Reviewer 2: Maybe...

```suggestion
Returns the initial data for a form field. It retrieves data from
:attr:`Form.initial` if present, otherwise trying :attr:`Field.initial`. 
Callable values are evaluated.
```

It is recommended to use :attr:`BoundField.initial` over
:meth:`~Form.get_initial_for_field()` because ``BoundField.initial`` has a
simpler interface. Also, unlike :meth:`~Form.get_initial_for_field()`,
:attr:`BoundField.initial` caches its values. This is useful especially when
dealing with callables whose return values can change (e.g. ``datetime.now`` or
``uuid.uuid4``)::

        >>> import uuid
        >>> class UUIDCommentForm(CommentForm):
        ...     identifier = forms.UUIDField(initial=uuid.uuid4)
        >>> f = UUIDCommentForm()
        >>> print(f.get_initial_for_field(f.fields['identifier'], 'identifier'))
        972ca9e4-7bfe-4f5b-af7d-07b3aa306334
        >>> print(f.get_initial_for_field(f.fields['identifier'], 'identifier'))
        69699b65-5679-47fa-a518-3c9b872e5be8

>         # Using BoundField.initial, for comparison

Reviewer 3: ```suggestion
        >>> # Using BoundField.initial, for comparison
```

Author: 🦅 👁️

>         >>> import uuid
>         >>> class UUIDCommentForm(CommentForm):
>         ...     identifier = forms.UUIDField(initial=uuid.uuid4)
>         >>> f = UUIDCommentForm()
>         >>> print(f.get_initial_for_field(f.fields['identifier'], 'identifier'))
>         972ca9e4-7bfe-4f5b-af7d-07b3aa306334
>         >>> print(f.get_initial_for_field(f.fields['identifier'], 'identifier'))
>         69699b65-5679-47fa-a518-3c9b872e5be8
>         # Using BoundField.initial, for comparison
>         >>> print(f['identifier'].initial)
>         9c50c0f6-966f-475e-9e1f-fb591e5dadaf
>         >>> print(f['identifier'].initial)
>         9c50c0f6-966f-475e-9e1f-fb591e5dadaf

Reviewer 3: It's over-indented:
 ```suggestion
    >>> import uuid
    >>> class UUIDCommentForm(CommentForm):
    ...     identifier = forms.UUIDField(initial=uuid.uuid4)
    >>> f = UUIDCommentForm()
    >>> print(f.get_initial_for_field(f.fields['identifier'], 'identifier'))
    972ca9e4-7bfe-4f5b-af7d-07b3aa306334
    >>> print(f.get_initial_for_field(f.fields['identifier'], 'identifier'))
    69699b65-5679-47fa-a518-3c9b872e5be8
    # Using BoundField.initial, for comparison
    >>> print(f['identifier'].initial)
    9c50c0f6-966f-475e-9e1f-fb591e5dadaf
    >>> print(f['identifier'].initial)
    9c50c0f6-966f-475e-9e1f-fb591e5dadaf
```

Reviewer 3: I would also remove `print()` from both examples. It's worth showing that `initial` returns `UUID` and `datetime` objects.
```python
        >>> f['identifier'].initial
        UUID('9c50c0f6-966f-475e-9e1f-fb591e5dadaf')
```

        <label for="myFIELD">...</label><input id="myFIELD" type="text" name="my_field" required>

.. attribute:: BoundField.initial

    Use :attr:`BoundField.initial` to retrieve initial data for a form field.
    It retrieves the value from :attr:`Form.initial`, or :attr:`Field.initial`
    if not present, and then evaluates if a callable. The return value is
    cached::

        >>> import uuid
        >>> class UUIDCommentForm(CommentForm):
        ...     identifier = forms.UUIDField(initial=uuid.uuid4)
        >>> f = UUIDCommentForm()
        >>> print(f['identifier'].initial)
        9c50c0f6-966f-475e-9e1f-fb591e5dadaf
        >>> print(f['identifier'].initial)
        9c50c0f6-966f-475e-9e1f-fb591e5dadaf

>     Using :attr:`BoundField.initial` is recommended over

Reviewer: I'm not sure we need the alternative showing that the non cached version changes. 🤔

Otherwise maybe chop the words but add the code examples to the above section with a code comment?

Author: As yet another alternative, this could be included under `Form.get_initial_for_field()`'s docs, but it would require defining `UUIDCommentForm` in two places.

Author: PS - I also like these `get_initial_for_field()` code examples because they show concretely how the `BoundField.initial` is simpler. :)

Author: I resolved this by putting the `get_initial_for_field()` code in the `get_initial_for_field()` section, and limiting mention of `get_initial_for_field()` in `BoundField.initial`'s to a brief sentence at the end. This minimizes the extraneous info when reading the "happy path" of the docs, while still acknowledging `get_initial_for_field()` for people that have code that can be improved, and wouldn't otherwise come across this info.

        <label for="myFIELD">...</label><input id="myFIELD" type="text" name="my_field" required>

.. attribute:: BoundField.initial

    Use :attr:`BoundField.initial` to retrieve initial data for a form field.

>     It retrieves the data from :attr:`Form.initial` if present, otherwise
>     :attr:`Field.initial`, and evaluates if a callable. See

Reviewer 2: Maybe... 

```suggestion
    It retrieves the data from :attr:`Form.initial` if present, otherwise
    trying :attr:`Field.initial`. Callable values are evaluated. See
```
